File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        
        def stringBuilder(o, c, s):
            if o == n and c == n:
                result.append(s)
            elif o == n:
                stringBuilder(o, c + 1, s + ")")
            elif o > c:
                stringBuilder(o + 1, c, s + "(")
                stringBuilder(o, c + 1, s + ")")
            else:
                stringBuilder(o + 1, c, s + "(")
        
        stringBuilder(0, 0, "")
        return result

File: Data Structures & Algorithms/non-cyclical-number/submission-0.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();

        int result = n;

        while (true) {
            result = sum(result);

            if (result == 1) return true;
            else {
                if (seen.contains(result)) return false;
                seen.add(result);
            }
        }
    }

    private int sum(int n) {
        int total = 0;

        while (n != 0) {
            int x = n % 10;
            x *= x;
            total += x;
            n /= 10;
        }

        return total;
    }
}
